x-pack/plugins/alerts/server/alerts_client.ts

```diff
@@ -614,13 +620,15 @@ export class AlertsClient {
       attributes = alert.attributes;
       version = alert.version;
     }
-
-    await this.authorization.ensureAuthorized(
-      attributes.alertTypeId,
-      attributes.consumer,
-      WriteOperations.Disable
-    );
-
+    // We only disable this whenever we know we have an API key that cannot pass
+    // this check and we are needing to disable the alert altogether
+    if (disableEnsureAuthorized !== true) {
+      await this.authorization.ensureAuthorized(
+        attributes.alertTypeId,
+        attributes.consumer,
+        WriteOperations.Disable
+      );
+    }
```

Comment:
A couple of thoughts that this triggers for me are:
1. I don't think we ever want to bypass the authorization check based on a user provided argument. I understand why you did this, but this feels like a hack intended to avoid solving this at a framework level.
2. I don't think this should be done at an `operation` level at all, but rather at framework level, where any error caused by an inability to decrypt the alert leads to it being disabled and this shouldn't be done by the AlertsClient itself (which already has too many responsibilities as it is) but rather a part of the framework which bypasses the AlertsClient and is called by the Task Runner.

Reply:
This ticket is good for discussions and see you're already there:
https://github.com/elastic/kibana/issues/74393

This PR is in draft mode not PR mode just yet, so it can work as a discussion point as well. I think draft mode works well for starting discussions from time to time which is why I enjoy doing them.

I started the code out within the task manager and only called this piece of code as it looked fairly natural. I can rework this code to not call into alerting client and just keep it within task_runner if that makes more sense to you.

I am also very open to pointers or tips on where we want to add this that benefits everyone as well. Just let me know. No problems if you have a better solution in mind that works for us? I just don't know what your team's capacity is at the moment which is why I wanted to put some code out here to start talks.

On our side we are getting reports of corrupted API keys from various scenarios during testing and migrations and people are wanting the alerts to be deactivated if it does happen regardless of where the bugs are originating which can be due to various situations as outlined in: https://github.com/elastic/kibana/issues/74393

Thanks for taking a look though @gmmorris

Reply:
> No problems if you have a better solution in mind that works for us?

I don't have a clear solution yet, but there are a few things we've been discussing around this, as this already came up in the context of migrations.
I want to make sure whichever solution we go with, it addresses the different cases we have:
1. What happens when decryption fails in migration?
2. What happens when it fails due to an operation in the AlertsClient?
3. What happens when it fails as part of a task run?
4. Do we want to actually disable at Saved Object or make it a "logical" disable, so that we can differentiate between an Alerts that was disabled on purpose and an Alert disabled due to an inability to decrypt?
5. How do we notify the user about this? We *definitely* need this to appear in the Event Log, so just disabling isn't enough. Do we need notify the producer? Do we need to notify the consumer? How best to do this?

There's probably more to it than just those things.
That's what I meant by this being broader than just what we need to do for siem, and my fear is that if we jump into a solution the doesn't take all these into account, we'll lock ourselves into something we can't change until a major version or even worse, something that can't address the needs of the other solutions and teams that are building on top of the framework.

Sorry if for the brain dump, I just wanted to share the wide range of concerns we're responsible for considering whenever we pick up an issue like this.

> I just don't know what your team's capacity is at the moment which is why I wanted to put some code out here to start talks.

That's fair, and we're happy to collaborate and approve changes in our code when it makes sense.
But I can assure that in this case, this issue will be addressed as part of 7.10, because it was already blocking the 7.10 release following the RBAC work which relies on migrations and needs to handle this case as well: https://github.com/elastic/kibana/issues/74404

If you're saying you need a solution in 7.9 then we can have a conversation about patching it somehow outside of the framework, but I definitely don't want to rush a solution into 7.9 (definitely not this late in the release cycle).

Reply:
> If we've introduced a regression in 7.9 then they need to know about it. If it isn't a regression, but rather a limitation in the current implementation, then we need to have a stack level discussion about addressing it. 

Feel free to loop us in any discussion if you need our input on ESO matters! Subscribed to https://github.com/elastic/kibana/issues/74393 for now.

Side note: I noticed PR uses plain try/catch, but if you decide to handle decryption errors make sure you rely on public `isEncryptionError` API exposed by the ESO plugin. We likely don't want to handle this _expected_ type of errors (these may potentially be handled at the alerting framework level) in the same way as we would handle unexpected failures (these likely should be propagated to the alerting consumers).

Reply:
If I've followed along correctly, then I think I would classify this as a limitation/expected behavior, rather than a regression.

The [`getDecryptedAsInternalUser`](https://github.com/elastic/kibana/blob/692db4f1725637194a525ef88b033cc658d2700a/x-pack/plugins/encrypted_saved_objects/server/saved_objects/index.ts#L75) function that the ESO plugin provides is designed to throw an [`EncryptionError`](https://github.com/elastic/kibana/blob/65186b3393b6c58c020226d8fbadf84e02c0e4a0/x-pack/plugins/encrypted_saved_objects/server/crypto/encryption_error.ts#L15) whenever we fail to decrypt a saved object. This could be due to an invalid decryption key, corrupt AAD, or some unexpected condition.

From my perspective, it seems that an invalid API key is just one of potentially many reasons why an alert may not be in a usable state. Today, the API Key is the most common/visible reason, but I'm hesitant to build something specifically for this scenario when there are likely others that the alerting framework will need to consider.